### PR TITLE
Add elastic net alpha to hyperparameter tuning vector

### PR DESCRIPTION
**Pull request by joshvfleming**

### CODE REVIEW

File: photon-client/src/test/scala/com/linkedin/photon/ml/io/scopt/game/ScoptGameTrainingParametersParserTest.scala [hunk 1]
```diff
@@ -164,7 +164,6 @@ class ScoptGameTrainingParametersParserTest {
       .put(GameTrainingDriver.treeAggregateDepth, treeAggregateDepth)
       .put(GameTrainingDriver.hyperParameterTuning, hyperparameterTuningMode)
       .put(GameTrainingDriver.hyperParameterTuningIter, hyperparameterTuningIter)
-      .put(GameTrainingDriver.hyperParameterTuningRange, hyperparameterTuningRange)
       .put(GameTrainingDriver.computeVariance, computeVariance)
```
Comment (ashelkovnykov) [inline photon-client/src/test/scala/com/linkedin/photon/ml/io/scopt/game/ScoptGameTrainingParametersParserTest.scala]: Should add custom ranges and alpha to this test

File: photon-lib/src/main/scala/com/linkedin/photon/ml/util/DoubleRange.scala [hunk 2]
```diff
@@ -23,4 +23,12 @@ package com.linkedin.photon.ml.util
...
 
   require(start < end, s"Invalid range: start $start comes after end date $end.")
+
+  /**
+   * Applies the transformation function to the range
+   *
+   * @param fn The transformation function
+   * @return The transformed range
+   */
+  def transform(fn: Double => Double): DoubleRange = DoubleRange(fn(start), fn(end))
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/util/DoubleRange.scala]: Should add a unit test for this

File: photon-client/src/test/scala/com/linkedin/photon/ml/io/scopt/game/ScoptGameTrainingParametersParserTest.scala (hunk 1 above)
Comment (joshvfleming) [reply photon-client/src/test/scala/com/linkedin/photon/ml/io/scopt/game/ScoptGameTrainingParametersParserTest.scala]: Done.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/util/DoubleRange.scala (hunk 2 above)
Comment (joshvfleming) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/util/DoubleRange.scala]: Done.

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala [hunk 3]
```diff
@@ -22,8 +22,11 @@ import org.apache.spark.sql.DataFrame
 
 import com.linkedin.photon.ml.estimators.GameEstimator.{GameOptimizationConfiguration, GameResult}
 import com.linkedin.photon.ml.hyperparameter.{EvaluationFunction, VectorRescaling}
-import com.linkedin.photon.ml.optimization.game.{FixedEffectOptimizationConfiguration, RandomEffectOptimizationConfiguration}
+import com.linkedin.photon.ml.hyperparameter.EvaluationFunction
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: Double import of `EvaluationFunction`

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala [hunk 4]
```diff
@@ -33,21 +36,44 @@ import com.linkedin.photon.ml.util.DoubleRange
...
+    validationData: DataFrame)
   extends EvaluationFunction[GameResult] {
 
+  import GameEstimatorEvaluationFunction._
+
   // CoordinateOptimizationConfigurations sorted in order by coordinate ID name
   private val baseConfigSeq = baseConfig.toSeq.sortBy(_._1)
 
+  // Pull the hyperparameter ranges from the optimization configuration
+  protected[estimators] val ranges = baseConfigSeq
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: Public/protected value requires explicit type annotation (Scala style)

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala [hunk 5]
```diff
@@ -33,21 +36,44 @@ import com.linkedin.photon.ml.util.DoubleRange
...
+    .flatMap {
+      case (_, config: GLMOptimizationConfiguration) => {
+        val regularizationWeightRange = config.regularizationWeightRange
+          .getOrElse(DEFAULT_REG_WEIGHT_RANGE)
+          .transform(log)
+
+        val elasticNetParamRange = config.elasticNetParamRange
+          .getOrElse(DEFAULT_REG_ALPHA_RANGE)
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: Fits on one line

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala [hunk 6]
```diff
@@ -33,21 +36,44 @@ import com.linkedin.photon.ml.util.DoubleRange
...
   // CoordinateOptimizationConfigurations sorted in order by coordinate ID name
   private val baseConfigSeq = baseConfig.toSeq.sortBy(_._1)
 
+  // Pull the hyperparameter ranges from the optimization configuration
+  protected[estimators] val ranges = baseConfigSeq
+    .flatMap {
+      case (_, config: GLMOptimizationConfiguration) => {
+        val regularizationWeightRange = config.regularizationWeightRange
+          .getOrElse(DEFAULT_REG_WEIGHT_RANGE)
+          .transform(log)
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: `.regularizationWeightRange` should also be on its own line

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala [hunk 7]
```diff
@@ -33,21 +36,44 @@ import com.linkedin.photon.ml.util.DoubleRange
...
+            Seq(regularizationWeightRange)
+        }
+      }
+
+      case _ => Seq()
+    }
+
   // Number of parameters in the base configuration
-  val numParams: Int = baseConfigSeq.length
+  val numParams = ranges.length
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: Public/protected value requires explicit type annotation (Scala style)

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala [hunk 8]
```diff
@@ -33,21 +36,44 @@ import com.linkedin.photon.ml.util.DoubleRange
...
 
+  import GameEstimatorEvaluationFunction._
+
   // CoordinateOptimizationConfigurations sorted in order by coordinate ID name
   private val baseConfigSeq = baseConfig.toSeq.sortBy(_._1)
 
+  // Pull the hyperparameter ranges from the optimization configuration
+  protected[estimators] val ranges = baseConfigSeq
+    .flatMap {
+      case (_, config: GLMOptimizationConfiguration) => {
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: Redundant braces for this case statement

File: photon-client/src/test/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunctionTest.scala [hunk 9]
```diff
@@ -35,16 +35,41 @@ import com.linkedin.photon.ml.util.DoubleRange
  * Unit tests for [[GameEstimatorEvaluationFunction]].
  */
 class GameEstimatorEvaluationFunctionTest {
+  import GameEstimatorEvaluationFunction._
```
Comment (ashelkovnykov) [inline photon-client/src/test/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunctionTest.scala]: Need space between lines 37 and 38

File: photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala (hunk 5 above)
Comment (joshvfleming) [reply photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: I fixed all the other ones, but I don't want to fix this one. If I put this all on one line, it doesn't match the initialization of `regularizationWeightRange` on L55 stylistically and looks super weird and IMO harder to read.

Comment (ashelkovnykov) [reply photon-client/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimatorEvaluationFunction.scala]: 10,000 years gulag

**Pull request merged**